core/lib/dal/src/tee_proof_generation_dal.rs

```diff
@@ -64,72 +64,93 @@ impl TeeProofGenerationDal<'_, '_> {
     ) -> DalResult<Option<LockedBatch>> {
         let processing_timeout = pg_interval_from_duration(processing_timeout);
         let min_batch_number = i64::from(min_batch_number.0);
+        let mut transaction = self.storage.start_transaction().await?;
+        let batch_number = sqlx::query!(
+            r#"
+            SELECT
+                p.l1_batch_number
+            FROM
+                proof_generation_details p
+            LEFT JOIN
+                tee_proof_generation_details tee
+                ON
+                    p.l1_batch_number = tee.l1_batch_number
+                    AND tee.tee_type = $1
+            WHERE
+                (
+                    p.l1_batch_number >= $5
+                    AND p.vm_run_data_blob_url IS NOT NULL
+                    AND p.proof_gen_data_blob_url IS NOT NULL
+                )
+                AND (
+                    tee.l1_batch_number IS NULL
+                    OR (
+                        (tee.status = $2 OR tee.status = $3)
+                        AND tee.prover_taken_at < NOW() - $4::INTERVAL
+                    )
+                )
+            LIMIT 1
+            FOR UPDATE OF p
+            SKIP LOCKED
```

Comment:
I had to explicitly use `OF p` here. If you leave it out or switch it to `OF tee`, you will run into this error:
```
error: error returned from database: FOR UPDATE cannot be applied to the nullable side of an outer join
```

Reply:
The error is understandable. Just to clarify, since it's the `tee_proof_generation_details` table that is updated below: Do I understand correctly that this new query can actually have `tee_proof_generation_details` side of the join absent (I guess if the query is invoked for a certain batch for the first time?). If so, I'd suggest to add a comment explaining why rows are locked on a table that's not modified in the transaction.

Reply:
> (...) Do I understand correctly that this new query can actually have tee_proof_generation_details side of the join absent (I guess if the query is invoked for a certain batch for the first time?). (...)

Yup, exactly.

> (...) If so, I'd suggest to add a comment explaining why rows are locked on a table that's not modified in the transaction.

Fixed here:
https://github.com/matter-labs/zksync-era/blob/ecd8c335272f6d80ca017f521c5282c469490168/core/lib/dal/src/tee_proof_generation_dal.rs#L69-L72

Lemme know how you would rephrase it if you feel it’s not accurate enough.